exercises/tictactoe/descr.md

```diff
@@ -0,0 +1,297 @@
+parameters `w` and `h` are the width and height of the board, while the
+parameter `k` is the size of an alignment that a player must achieve in order
+to win. The parameters `w`, `h`, `k` remain fixed throughout an entire game.
+In the game of Tic-Tac-Toe, they are equal to 3.
+
+We represent a game state in OCaml as a record with the following fields:
+
+```
+  type state = {
+    w: int; h: int; k: int;
+    player_bitmap: bitmap;
+    opponent_bitmap: bitmap;
+    past_moves: int;
+  }
+```
+
+The current state of the board is represented by *two bitmaps*. The bitmap
+`player_bitmap` stores the marks of the player who is currently up, while the
+bitmap `opponent_bitmap` stores the marks of his opponent.
+
+The integer field `past_moves` stores the number of moves that have been
+played. This field is redundant, as its value can be recovered by counting how
+many bits are set in the bitmaps `player_bitmap` and `opponent_bitmap`.
+Nevertheless, it is convenient to have this field: this removes the need for
+an efficient implementation of the "bit count" operation.
+
+It is now up to you to write a number of simple auxiliary functions that
+operate on game states, namely `initial`, `square_is_available`, `play`,
+`no_square_is_available`, and `opponent_has_won`. (No automatic grading
+is performed at this stage.) Then, you can move on to the final question:
+
+**Question 8.** Define a function `tree` of type `state -> tree`
+such that `tree state` is the game tree associated with
+the game state `state`.
+
+*Note.* The definition of a game tree is explained in the
+<a href="" onclick="top.location='/exercises/alpha_beta/';">companion
+exercise</a>
+on the Alpha-Beta algorithm.
+
+*Hint.* In a game tree, the value of a leaf is always relative to the player
+who is currently up. This value should be +1 if this player has won, 0 if the
+game is a draw, and -1 if this player has lost.
+
+*Hint.* In Tic-Tac-Toe, as in many common games, there cannot be a leaf whose
+value is +1. That would mean that our opponent has just played, yet we have
+won. This is impossible: no move can cause a player to lose immediately.
+
+*Hint.* The functions `Seq.interval`, `Seq.filter`, and `Seq.map` can be
+ helpful in the construction of a sequence of permitted moves.
+
+## To Go Further
+
+If you have completed both this exercise and its <a href=""
+onclick="top.location='/exercises/alpha_beta/';">companion exercise</a> on the
+Alpha-Beta algorithm,
+then we recommend combining them to solve Tic-Tac-Toe.
+You can verify that, if both players play perfectly,
+then the game must end in a draw.
+Although the game tree for Tic-Tac-Toe has size 549946,
+the Alpha-Beta algorithm evaluates it in a fraction of a second.
+(The algorithm builds and explores only part of the tree.)
+
+To explore larger values of `w`, `h`, and `k`, such as (4, 4, 4),
+it is necessary to improve the Alpha-Beta algorithm; in particular,
+a transposition table is required.
```

Comment:
This sentence is a bit mysterious to me. What kind of "transposition table" are you referring to?

Reply:
A "transposition table" is standard terminology (although admittedly somewhat obscure) in the literature on search algorithms; it is just a hash table that records the game configurations that have already been seen, possibly up to symmetries ("transpositions").

We could improve the text to make it clear that this is a hash table and/or add a link such as https://en.wikipedia.org/wiki/Transposition_table .

Reply:
```suggestion
a [transposition table](https://en.wikipedia.org/wiki/Transposition_table) is required.
```